server/teamshandler.go

```diff
@@ -0,0 +1,133 @@
+
+	// [ ] decode request JSON
+	// [ ] validate we have a public key referred to in the detached signature
+	// [ ] validate the detached signature of the team roster
+	// [ ] validate the team UUID is non-zero
+	// [ ] ensure we don't have the existing team UUID already (create only atm)
+	// [ ] validate the signing key is listed against a member in the roster
+	// [ ] ... and that the email of that member is verified in the API
+	// [ ] save the team roster, signature and created_at
+	// [ ] return HTTP status created
+
+	requestData := v1structs.UpsertTeamRequest{}
+	if err := decodeJsonRequest(r, &requestData); err != nil {
+		writeJsonError(w, err, http.StatusBadRequest)
+		return
+	}
+
+	if requestData.TeamRoster == "" {
+		writeJsonError(w, fmt.Errorf("missing teamRoster"), http.StatusBadRequest)
+		return
+	}
+
+	if requestData.ArmoredDetachedSignature == "" {
+		writeJsonError(w, fmt.Errorf("missing armoredDetachedSignature"), http.StatusBadRequest)
+		return
+	}
+
+	apparentSignerKey, err := getAuthorizedUserPublicKey(r)
+	if err == errAuthKeyNotFound {
+		writeJsonError(w,
+			fmt.Errorf("public key that signed the roster has not been uploaded"),
+			http.StatusBadRequest)
+		return
+	} else if err != nil {
+		writeJsonError(w, err, http.StatusBadRequest)
+		return
+	}
+
+	if err = checkRosterSignature(requestData, apparentSignerKey); err != nil {
+		log.Printf("roster signature verification failed: %v", err)
+		writeJsonError(w, fmt.Errorf("signature verification failed"), http.StatusBadRequest)
+		return
+	}
+
+	team, err := team.Parse(strings.NewReader(requestData.TeamRoster))
+	if err != nil {
+		writeJsonError(w, err, http.StatusBadRequest)
+		return
+	}
+
+	if err := team.Validate(); err != nil {
+		writeJsonError(w, err, http.StatusBadRequest)
+		return
+	}
+
+	person, err := team.GetPersonForFingerprint(apparentSignerKey.Fingerprint())
+	if err != nil {
+		writeJsonError(w,
+			fmt.Errorf("signing key's fingerprint isn't listed in roster"), http.StatusBadRequest)
+		return
+	}
+
+	err = datastore.RunInTransaction(func(txn *sql.Tx) error {
+
+		if verified, err := datastore.QueryEmailVerifiedForFingerprint(
+			txn, person.Email, apparentSignerKey.Fingerprint()); err != nil {
+
+			return fmt.Errorf("error querying email verification: %v", err)
+		} else if !verified {
+
+			return fmt.Errorf("signing key's email listed in roster is unverified")
+		}
+
+		if exists, err := datastore.TeamExists(txn, team.UUID); err != nil {
+			return fmt.Errorf("error querying for team: %v", err)
+		} else if exists {
+			return fmt.Errorf("team with UUID %s already exists", team.UUID)
+		}
+
+		team := datastore.Team{
+			UUID:            team.UUID,
+			Roster:          requestData.TeamRoster,
+			RosterSignature: requestData.ArmoredDetachedSignature,
+			CreatedAt:       time.Now(),
+		}
+
+		if err := datastore.CreateTeam(txn, team); err != nil {
+			return fmt.Errorf("error creating team: %v", err)
+		}
+
+		return nil
```

Comment:
You don't specify `nil` elsewhere?

Reply:
Did you spot that we're inside a function that's being passed into `RunInTransaction` ?

Reply:
Ah no, I missed that, thanks!